RFCs/0016-Loggers.md

```diff
@@ -0,0 +1,85 @@
+# 0016 Loggers
+
+## Summary
+This note outlines the proposed changes for:
+1. Enabling logger support in test platform for TranslationLayer clients.
+2. Allowing users to provide loggers from the runsettings.
+
+## Motivation
+1. Enabling logger support for TranslationLayer clients will allow clients to use logger extensibility feature.
+
+2. Allowing users to provide loggers from the runsettings enables:
+  * TranslationLayer clients to provide loggers to the test platform.
+  * Alternative way to provide loggers in addition to the existing command line logger argument.
+
+## Specifying a logger
+Loggers can be provided to the test platform in one of the following ways:
+
+1.  "/Logger:"UriOrFriendlyName";Key1=Value1;Key2=Value2" This is a switch to vstest.console.exe that feeds logger information to the test platform. For instance:
+
+    ```
+    /Logger:"logger://Microsoft/TestPlatform/TrxLogger/v1"
+    ```
+
+    This loads and initializes logger with "logger://Microsoft/TestPlatform/TrxLogger/v1" as uri.
+
+    ```
+    /Logger:sampleLogger;Key1=Value1;Key1=Value2
+    ```
+
+    This loads and initializes logger with "sampleLogger" as friendly name. Key1=Value1 and Key2=Value2 are passed as dictionary parameters to the logger while initialization.
+
+2. Runsettings via "Logger" node in the LoggerRunSettings section. Here is a sample on how this can be specified:
+
+    ```xml
+    <RunSettings>
+        <LoggerRunSettings>
+            <Loggers>
+                <Logger friendlyName="sampleLoggerwithParameters">
+                    <Configuration>
+                        <Key1>Value1</Key1>
+                        <Key2>Value2</Key2>
+                    </Configuration>
+                </Logger>
+                <Logger uri="logger://sample/sampleLoggerWithoutParameters1"
+                        friendlyName="sampleLoggerWithoutParameters1" />
+                <Logger uri="logger://sample/sampleLoggerWithoutParameters2"
+                        assemblyQualifiedName="Sample.Sample.Sample.SampleLogger, Sample.Sample.Logger, Version=0.0.0.0, Culture=neutral, PublicKeyToken=xxxxxxxxxxxxxxxx"
+                        friendlyName="sampleLoggerWithoutParameters2" />
+            </Loggers>
+        </LoggerRunSettings>
+    </RunSettings>
+    ```
+
+    This loads and initializes:
+    * Logger with friendlyName="sampleLoggerwithParameters". Key1=Value1 and Key2=Value2 are passed as dictionary parameters to the logger while initialization.
+    * Logger with uri="logger://sample/sampleLoggerWithoutParameters1". FriendlyName is ignored in this case as uri takes more precedence.
+    * Logger with assemblyQualifiedName="Sample.Sample.Sample.SampleLogger, Sample.Sample.Logger, Version=0.0.0.0, Culture=neutral, PublicKeyToken=xxxxxxxxxxxxxxxx". Uri and friendlyName are ignored in this case as assemblyQualifiedName takes more precedence.
```

Comment:
And user can override any runsetting with command line params like so `vstest.console foo.dll -- RunSettings.Loggers.Logger`. Is this supported? And will the precedence order be respected in this case?

Reply:
No, this is not supported.

Reply:
What's the rationale for not supporting it? At the least we shouldn't block this please. It will complicate implementation as I understand.

Reply:
Sorry for the confusion. User can override logger runsetting from command line paramas. I will update this info in RFC. thanks